### DISCUSSION TITLE
Use `$LDFLAGS` as default for `$ESMF_*LINKOPTS`

### DISCUSSION THREAD

**Discussion by xylar** [esmf-org/esmf#255] [Ideas / Requests]
On conda-forge, we recently ran into an issue when upgrading to OpenMPI 5 that ESMF tests were failing with a linking error:
```
mpif90   -m64 -mcmodel=small -pthread -Wl,--no-as-needed -fopenmp -L$SRC_DIR/lib/libO/Linux.gfortran.64.openmpi.default -L$PREFIX/lib -L$PREFIX/lib -L$BUILD_PREFIX/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/ -Wl,-rpath,$SRC_DIR/lib/libO/Linux.gfortran.64.openmpi.default -Wl,-rpath,$PREFIX/lib  -Wl,-rpath,$PREFIX/lib -Wl,-rpath,$BUILD_PREFIX/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/ -o $SRC_DIR/test/testO/Linux.gfortran.64.openmpi.default/ESMF_StringUTest ESMCI_StringSubr.o  ESMF_StringUTest.o -lesmf  -lrt -lstdc++ -ldl -lnetcdff -lnetcdf -lpioc
/home/conda/feedstock_root/build_artifacts/esmf_1717757906123/_build_env/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/conda/feedstock_root/build_artifacts/esmf_1717757906123/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/./libpmix.so.2: undefined reference to `clock_gettime@GLIBC_2.17'
/home/conda/feedstock_root/build_artifacts/esmf_1717757906123/_build_env/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/conda/feedstock_root/build_artifacts/esmf_1717757906123/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/libmpi_mpifh.so: undefined reference to `memcpy@GLIBC_2.14'
collect2: error: ld returned 1 exit status
```

After some lengthy discussion, it was discovered that the issue was that `$LDFLAGS` was not being passed along to the linker by default, and setting:
```
export ESMF_CLINKOPTS="$LDFLAGS"
export ESMF_F90LINKOPTS="$LDFLAGS"
export ESMF_CXXLINKOPTS="$LDFLAGS"
```
fixed the issue.  The feeling from a conda-forge maintainer (see https://github.com/conda-forge/esmf-feedstock/pull/117/files#r1634670892) was that this should probably be the default value within ESMF, so I'm proposing this change.

**Comment by danrosen25** [comment]
@theurich @oehmke 
Does this seem system specific as it has to do with the GNU C library not being linked by the linker?
Or are we missing something in the build recipe for the ESMF_StringUTest executable?
```
mpif90
-m64
-mcmodel=small
-pthread
-Wl,--no-as-needed
-fopenmp
-L$SRC_DIR/lib/libO/Linux.gfortran.64.openmpi.default
-L$PREFIX/lib
-L$PREFIX/lib
-L$BUILD_PREFIX/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/
-Wl,-rpath,$SRC_DIR/lib/libO/Linux.gfortran.64.openmpi.default
-Wl,-rpath,$PREFIX/lib
-Wl,-rpath,$PREFIX/lib
-Wl,-rpath,$BUILD_PREFIX/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/
-o $SRC_DIR/test/testO/Linux.gfortran.64.openmpi.default/ESMF_StringUTest
ESMCI_StringSubr.o
ESMF_StringUTest.o
-lesmf
-lrt
-lstdc++
-ldl
-lnetcdff
-lnetcdf
-lpioc
```

**Comment by oehmke** [comment]
I think that this is probably a question for @theurich <https://github.com/theurich>, if it’s ok to wait until he has a chance to answer.

**Comment by danrosen25** [comment]
@theurich 
Is this something we'd like to add to the backlog? As of now we don't pick up standard compiler/linker configuration options that are commonly picked up from the environment by other build systems.

- CC
- CFLAGS
- CXX
- CXXFLAGS
- FC
- FFLAGS
- LDFLAGS

**Comment by danrosen25** [comment]
ESMF intentionally separates ESMF build configuration from the common user environment variables (i.e. CC, FC, LDFLAGS) using the ESMF_ prefix in order to make sure all configuration settings are explicitly set. Although it's a pain point for more experienced users it prevents less experienced users from mistakenly using unintentional build settings. We will update the Quick Start Build in the User's Guide with some clearer build instructions for using common user environment variables.
https://earthsystemmodeling.org/docs/release/latest/ESMF_usrdoc/
